### ISSUE TITLE
Add new compiler configuration with bound checks in std-container and `AddressSanitizer` for array bounds checks

### ISSUE THREAD

**Issue by DimitriPlotnikov** [nest/nest-simulator#756]
Std-containers have safe (with bound checks, e.g. vec.at(1)) and unsafe (without bounds checks vec[1]) getters. However, `g++` enables to use a "debug" version of these containers by defining a flag `D_GLIBCXX_DEBUG` for the g++ compiler. In the later case, all unsafe access will be replaces through the safe variants.

The same applies for the plain c++ arrays. By default, the g++ compiler relays on the program correctness and omits any checks during runtime. However, g++ can do them if the `-fsanitize=bounds` flag is passed to the compiler.

I've created a small demonstrator program (see. attachment.) and tested it with the `gcc` 6.3.0
Uncommenting line 10  (which contains out-of-bound access to a vector) generates the following error:
```
/usr/include/c++/6/debug/vector:415:
Error: attempt to subscript container with out-of-bounds index -1, but 
container only holds 20 elements.

Objects involved in the operation:
    sequence "this" @ 0x0x7fff273c5930 {
      type = std::__debug::vector<char, std::allocator<char> >;
    }
```

Uncommenting line 12 generates the following error  (which contains out-of-bound access to an array:
```
test.cpp:12:26: runtime error: index 5 out of bounds for type 'int [5]'
```
[std_cpp_testing.zip](https://github.com/nest/nest-simulator/files/1084017/std_cpp_testing.zip)

**Comment by DimitriPlotnikov** [comment]
An pleasant effect in the previously described approach is that in the case of the std-container bound violation there are no actual information regarding the actual line of code where the error occurred. Therefore, I tested also `clang++-4.0`
Uncommenting line 10 (which contains out-of-bound access to a vector) generates the following error:
```
=================================================================
==14418==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000000f at pc 0x00000050cf75 bp 0x7ffd85d56cb0 sp 0x7ffd85d56ca8
READ of size 1 at 0x60300000000f thread T0
    #0 0x50cf74  (/home/nestml/Desktop/std_cpp_testing/test+0x50cf74)
    #1 0x7f596480d3f0  (/lib/x86_64-linux-gnu/libc.so.6+0x203f0)
    #2 0x419f49  (/home/nestml/Desktop/std_cpp_testing/test+0x419f49)

0x60300000000f is located 1 bytes to the left of 20-byte region [0x603000000010,0x603000000024)
allocated by thread T0 here:
    #0 0x5089a0  (/home/nestml/Desktop/std_cpp_testing/test+0x5089a0)
    #1 0x50d84b  (/home/nestml/Desktop/std_cpp_testing/test+0x50d84b)
    #2 0x50d7eb  (/home/nestml/Desktop/std_cpp_testing/test+0x50d7eb)
    #3 0x50d792  (/home/nestml/Desktop/std_cpp_testing/test+0x50d792)
    #4 0x50d64e  (/home/nestml/Desktop/std_cpp_testing/test+0x50d64e)
    #5 0x50d394  (/home/nestml/Desktop/std_cpp_testing/test+0x50d394)
    #6 0x50d1af  (/home/nestml/Desktop/std_cpp_testing/test+0x50d1af)
    #7 0x50cf15  (/home/nestml/Desktop/std_cpp_testing/test+0x50cf15)
    #8 0x7f596480d3f0  (/lib/x86_64-linux-gnu/libc.so.6+0x203f0)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/nestml/Desktop/std_cpp_testing/test+0x50cf74) 
Shadow bytes around the buggy address:
  0x0c067fff7fb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c067fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c067fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c067fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c067fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c067fff8000: fa[fa]00 00 04 fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8010: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==14418==ABORTING
```
Uncommenting line 12 (which contains out-of-bound access to an array) generates the following error:
```
==14454==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd9ae45074 at pc 0x00000050cf67 bp 0x7ffd9ae44fd0 sp 0x7ffd9ae44fc8
READ of size 4 at 0x7ffd9ae45074 thread T0
    #0 0x50cf66  (/home/nestml/Desktop/std_cpp_testing/test+0x50cf66)
    #1 0x7f407a5ca3f0  (/lib/x86_64-linux-gnu/libc.so.6+0x203f0)
    #2 0x419f19  (/home/nestml/Desktop/std_cpp_testing/test+0x419f19)

Address 0x7ffd9ae45074 is located in stack of thread T0 at offset 148 in frame
    #0 0x50cd6f  (/home/nestml/Desktop/std_cpp_testing/test+0x50cd6f)

  This frame has 4 object(s):
    [32, 56) 'testant1'
    [96, 97) 'ref.tmp'
    [112, 113) 'ref.tmp1'
    [128, 148) 'testant2' <== Memory access at offset 148 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (/home/nestml/Desktop/std_cpp_testing/test+0x50cf66) 
Shadow bytes around the buggy address:
  0x1000335c09b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c09c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c09d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c09e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c09f0: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
=>0x1000335c0a00: 00 00 00 f2 f2 f2 f2 f2 01 f2 01 f2 00 00[04]f3
  0x1000335c0a10: f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c0a20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c0a30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c0a40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000335c0a50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==14454==ABORTING
```

**Comment by DimitriPlotnikov** [comment]
Summary: with `gcc` an out-of-bound array access can be seen directly. With `clang` there is at least a chance for somebody how understands what to do with stacktraces to find out there the out-of-bound container access happened.

**Comment by heplesser** [comment]
@DimitriPlotnikov Thank you very much for figuring this out. I would suggest the following steps:
1. Add the necessary options to the `--Dwith-debug` option in [cmake/ProcessOptions.cmake](https://github.com/nest/nest-simulator/blob/372f485d036e5e11ac7619abce8b4429cbdf13af/cmake/ProcessOptions.cmake#L35), so that the compiler checks access violations for STL containers and C-style arrays. We may need to check here that the compiler accepts these switches.
1. Modify the Travis setup to run in debug. I would think that it makes most sense to compile NEST debug mode for all test configurations. @lekshmideepu, what do you think?

**Comment by heplesser** [comment]
At least on my machine, both compiling and running NEST with both debugging options given by @DimitriPlotnikov becomes very slow. Thus, it is probably not a good idea to use this for all tests, but rather for a test that is only run after all normal configurations have passed.

**Comment by terhorstd** [comment]
this is currently blocked by #789

**Comment by terhorstd** [comment]
This would be nice to have checked in one run of the travis build.

**Comment by terhorstd** [comment]
This is something that should be tested in CI, but because of its computational cost, depends on stages in CI. It doesn't make sense to run this each and every time, but only occasionally and if everything else passes.

@lekshmideepu this is a good use-case for improved travis setup.

**Comment by terhorstd** [comment]
Since this would more be a debug strategy than a test that would need to run continuously (especially since it causes longish run times), we decided to close this after adding some notes in the developer documentation.

**Comment by heplesser** [comment]
#2169 adds `-D_GLIBCXX_ASSERTIONS` to all test build configurations, catching all misaligned STL container access in binaries built with g++. Since we use practically no pure C arrays in NEST, I consider this good enough and close this issue.
